### PR TITLE
GUVNOR-2173 - Asset Oracle: Change Impact prediction

### PR DESCRIPTION
**Pull request by mrietveld**
Change Impact, Part I: building a reference infrastructure that we can query.

For an explanation, see https://github.com/droolsjbpm/kie-wb-common/pull/414

PR set:
- https://github.com/uberfire/uberfire/pull/455
- https://github.com/droolsjbpm/kie-wb-common/pull/414
- https://github.com/droolsjbpm/drools-wb/pull/257
- https://github.com/droolsjbpm/jbpm-designer/pull/375
- https://github.com/droolsjbpm/jbpm-form-modeler/pull/63

### CODE REVIEW

File: uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-lucene/src/main/java/org/uberfire/ext/metadata/backend/lucene/index/CustomAnalyzerWrapperFactory.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.uberfire.ext.metadata.backend.lucene.index;
+
+import java.util.Map;
+
+import org.apache.lucene.analysis.Analyzer;
+
+/**
```
Comment (manstis) [inline uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-lucene/src/main/java/org/uberfire/ext/metadata/backend/lucene/index/CustomAnalyzerWrapperFactory.java]: Liking empty JavaDoc blocks less than no JavaDocs...

**Comment by manstis** [comment]
Personally I'd like to see the empty Javadoc block removed or embellished!

**Comment by ederign** [comment]
@mrietveld could you please apply @manstis suggestion?

**Comment by mrietveld** [comment]
@manstis, @ederign: fixed (added javadoc). Thanks!

File: uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-lucene/src/main/java/org/uberfire/ext/metadata/backend/lucene/index/CustomAnalyzerWrapperFactory.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import org.apache.lucene.analysis.Analyzer;
+import org.uberfire.ext.metadata.backend.lucene.LuceneConfigBuilder;
+import org.uberfire.ext.metadata.backend.lucene.analyzer.FilenameAnalyzer;
+
+/**
+ * This factory allows us to extend the granularity used when configuring {@link Analyzer} instances.
+ * </p>
+ * For example, with normal configuration, {@link Analyzer}'s can only be applied to fields that start with a pre-defined prefix.
+ * <ul>
+ * <li>For example: use the {@link FilenameAnalyzer} for all fields with the key ""projectRoot"
```
Comment (manstis) [inline uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-lucene/src/main/java/org/uberfire/ext/metadata/backend/lucene/index/CustomAnalyzerWrapperFactory.java]: Double quotation mark (before ```projectRoot```)...

File: uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-lucene/src/main/java/org/uberfire/ext/metadata/backend/lucene/index/CustomAnalyzerWrapperFactory.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ * </ul>
+ * However, you can not use regular expressions here, which means that it is impossible to configure an analyzer for if have a
```
Comment (manstis) [inline uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-lucene/src/main/java/org/uberfire/ext/metadata/backend/lucene/index/CustomAnalyzerWrapperFactory.java]: Here or there? Surely the whole point of this interface is to support use of RegEx?

Comment (manstis) [inline uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-lucene/src/main/java/org/uberfire/ext/metadata/backend/lucene/index/CustomAnalyzerWrapperFactory.java]: Grammer: "...impossible to configure an analyzer **for if have a**..."?

File: uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-lucene/src/main/java/org/uberfire/ext/metadata/backend/lucene/index/CustomAnalyzerWrapperFactory.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * set of fields that have a pre-defined <em>prefix</em> but an dynamically-determined suffix (and we have that for change impact).
```
Comment (manstis) [inline uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-lucene/src/main/java/org/uberfire/ext/metadata/backend/lucene/index/CustomAnalyzerWrapperFactory.java]: "... but **a** dynamically.."

File: uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-lucene/src/main/java/org/uberfire/ext/metadata/backend/lucene/index/CustomAnalyzerWrapperFactory.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ * </p>
+ * This factory allows us to define our own {@link Analyzer} which can have it own logic to deal with this problem.
```
Comment (manstis) [inline uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-lucene/src/main/java/org/uberfire/ext/metadata/backend/lucene/index/CustomAnalyzerWrapperFactory.java]: "... have **its** own logic..."

**Comment by manstis** [comment]
Quality @mrietveld :-)

+1 from me (but I assume you want this to wait for the other PRs too)

**Comment by mrietveld** [comment]
This can be merged by itself, since it's not dependent on other PR's.

**Comment by ederign** [comment]
thanks @mrietveld

**Pull request merged**